### PR TITLE
Implemented FileStore for easy caching on disk

### PR DESCRIPTION
**Pull request by pietervdvn**
Hey Aliostad,

As you requested, a PR. "De koe bij de horens vatten", as we say in Dutch (To grasp the nettle; literally translated: to grasp the cow at its horns).

In summary: I implemented a simple cache to disk cachestore. There is _no_ automatic clearing of the cache (I hope that is handled somewhere in CacheCow and that CacheCow will call 'ClearCache' every now and then).

I've also added one test for the FileStore (which should cover all the code) and a bit of documentation into ICacheStore.

Furthermore, although I have quite some programming experience, I'm pretty new with C#. I might thus have made some errors considering Idioms - please point them out to me and I'll fix it.

### CODE REVIEW

File: src/CacheCow.Client.FileCacheStore/FileStore.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+        /// <inheritdoc />
+        public async Task<HttpResponseMessage> GetValueAsync(CacheKey key)
+        {
+            if (!File.Exists(_pathFor(key)))
+            {
+                return null;
+            }
+
+            var fs = File.OpenRead(_pathFor(key));
```
Comment (IDisposable) [inline src/CacheCow.Client.FileCacheStore/FileStore.cs]: Probably should wrap this in a `try` / `finally` or a `using` to ensure that if the `DeserializeToResponseAsync` throws we always Close/Dispose the file stream.

File: src/CacheCow.Client.FileCacheStore/FileStore.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            var resp = await _serializer.DeserializeToResponseAsync(fs);
+            fs.Close();
+            return resp;
+        }
+
+        /// <inheritdoc />
+        public async Task AddOrUpdateAsync(CacheKey key, HttpResponseMessage response)
+        {
+            var fs = File.OpenWrite(_pathFor(key));
```
Comment (IDisposable) [inline src/CacheCow.Client.FileCacheStore/FileStore.cs]: A `using` would be appropriate here too...

File: src/CacheCow.Client.FileCacheStore/FileStore.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            fs.Close();
+        }
+
+        /// <inheritdoc />
+        public Task<bool> TryRemoveAsync(CacheKey key)
+        {
+            return new Task<bool>(() =>
+                {
+                    if (!File.Exists(_pathFor(key)))
+
```
Comment (IDisposable) [inline src/CacheCow.Client.FileCacheStore/FileStore.cs]: _nit-pick_ extra line

File: src/CacheCow.Client.FileCacheStore/FileStore.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            foreach (var f in Directory.GetFiles(_cacheRoot))
+            {
+               File.Delete(f);
+            }
+        }
+
+
+        private string _pathFor(CacheKey key)
+        {
+            // Who the fuck thought using '/' in a Base64-encoding was a good idea?
```
Comment (IDisposable) [inline src/CacheCow.Client.FileCacheStore/FileStore.cs]: probably should not have that curse...

File: src/CacheCow.Client.FileCacheStore/FileStore.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            return _cacheRoot + "/" + key.HashBase64.Replace('/', '!');
+        }
+
+
```
Comment (IDisposable) [inline src/CacheCow.Client.FileCacheStore/FileStore.cs]: _nit-pick_ extra line

File: src/CacheCow.Client.FileCacheStore/FileStore.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                    File.Delete(_pathFor(key));
+                    return true;
+                }
+            );
+        }
+
+        /// <inheritdoc />
+        public async Task ClearAsync()
+        {
+            foreach (var f in Directory.GetFiles(_cacheRoot))
```
Comment (IDisposable) [inline src/CacheCow.Client.FileCacheStore/FileStore.cs]: I wonder if we should prevent foot-shooting and throw if the _cacheRoot hasn't been set to something non-blank... I can just imagine someone doing a `ClearAsync` and wiping the entire directory of a website...

**Comment by pietervdvn** [comment]
Alright, I'll look to it tomorrow

**Comment by pietervdvn** [comment]
I've taken your issues into account; please merge or let me know if there are still opportunities to improve.

**Comment by aliostad** [comment]
Thank you. It looks good so far, let me finish review and hopefully merge.

**Comment by pietervdvn** [comment]
Great! Thanks for merging.

Any idea when the NuGet-package will be updated?

File: src/CacheCow.Client.FileCacheStore/CacheCow.Client.FileCacheStore.csproj [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
```
Comment (aliostad) [inline src/CacheCow.Client.FileCacheStore/CacheCow.Client.FileCacheStore.csproj]: This should be both net452 and netstandard2.0 similar to other.

**Comment by pietervdvn** [comment]
I have edited the .csproj based on the .csproj of cachecow itself and did what seemed right. Please review carefully - I'm still very new to C# and have never really looked into these files before.

**Comment by aliostad** [comment]
Thank you @pietervdvn 

There are some areas where this could bomb out for example in case of concurrency when the same resources is accessed using multiple threads. But it is OK for now.

**Comment by aliostad** [comment]
It is on Nuget now:
https://www.nuget.org/packages/CacheCow.Client.FileStore/2.4.2

**Comment by pietervdvn** [comment]
Superb! It is integrated already in my code.

**Pull request merged**